File: backend-go/internal/service/accident_service.go
package service

import (
	"errors"
	"fmt"
	"time"

	"gorm.io/gorm"

	"github.com/instaagrammeta/rentacar/backend-go/internal/apperr"
	"github.com/instaagrammeta/rentacar/backend-go/internal/models"
)

// GetAccident returns an accident (with relations + photos) or a 404 error.
func (s *Service) GetAccident(id uint) (*models.Accident, error) {
	var a models.Accident
	if err := s.DB.Preload("Car").Preload("Client").Preload("Photos").First(&a, id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, apperr.NotFound(fmt.Sprintf("Accident #%d не найден", id))
		}
		return nil, err
	}
	return &a, nil
}

// ListAccidents returns accidents filtered by car.
func (s *Service) ListAccidents(carID uint, page, perPage, offset int) (*ListResult, error) {
	q := s.DB.Model(&models.Accident{})
	if carID > 0 {
		q = q.Where("car_id = ?", carID)
	}
	var total int64
	q.Count(&total)

	var items []models.Accident
	if err := q.Preload("Car").Preload("Client").Preload("Photos").
		Order("id desc").Limit(perPage).Offset(offset).Find(&items).Error; err != nil {
		return nil, err
	}
	out := make([]map[string]interface{}, 0, len(items))
	for i := range items {
		out = append(out, items[i].ToMap())
	}
	return &ListResult{Items: out, Total: total, Page: page, PerPage: perPage}, nil
}

// AccidentInput holds create/update fields for an accident.
type AccidentInput struct {
	CarID        *uint     `json:"car_id"`
	ClientID     *uint     `json:"client_id"`
	RentalID     *uint     `json:"rental_id"`
	AccidentDate *string   `json:"accident_date"`
	Description  *string   `json:"description"`
	RepairCost   *float64  `json:"repair_cost"`
	Photos       *[]string `json:"photos"`
}

// CreateAccident records an accident.
func (s *Service) CreateAccident(in AccidentInput, actor *Actor) (*models.Accident, error) {
	if in.CarID == nil {
		return nil, apperr.Validation("Автомобиль обязателен")
	}
	car, err := s.GetCar(*in.CarID)
	if err != nil {
		return nil, err
	}

	accDate := time.Now()
	if in.AccidentDate != nil && *in.AccidentDate != "" {
		if d, err := parseDate(*in.AccidentDate); err == nil && d != nil {
			accDate = *d
		}
	}

	accident := &models.Accident{
		CarID:        car.ID,
		ClientID:     in.ClientID,
		RentalID:     in.RentalID,
		AccidentDate: accDate,
		Description:  in.Description,
		RepairCost:   derefFloat(in.RepairCost),
	}
	if in.Photos != nil {
		for _, p := range *in.Photos {
			accident.Photos = append(accident.Photos, models.AccidentPhoto{FilePath: p})
		}
	}
	if err := s.DB.Create(accident).Error; err != nil {
		return nil, err
	}
	s.record(actor, "create_accident", "accident", uintPtr(accident.ID), "")
	return s.GetAccident(accident.ID)
}

// UpdateAccident updates an accident record.
func (s *Service) UpdateAccident(id uint, in AccidentInput, actor *Actor) (*models.Accident, error) {
	accident, err := s.GetAccident(id)
	if err != nil {
		return nil, err
	}
	if in.Description != nil {
		accident.Description = in.Description
	}
	if in.RepairCost != nil {
		accident.RepairCost = *in.RepairCost
	}
	if in.AccidentDate != nil && *in.AccidentDate != "" {
		if d, err := parseDate(*in.AccidentDate); err == nil && d != nil {
			accident.AccidentDate = *d
		}
	}
	if err := s.DB.Save(accident).Error; err != nil {
		return nil, err
	}
	if in.Photos != nil {
		s.DB.Where("accident_id = ?", accident.ID).Delete(&models.AccidentPhoto{})
		for _, p := range *in.Photos {
			s.DB.Create(&models.AccidentPhoto{AccidentID: accident.ID, FilePath: p})
		}
	}
	return s.GetAccident(accident.ID)
}

// DeleteAccident deletes an accident record.
func (s *Service) DeleteAccident(id uint, actor *Actor) error {
	accident, err := s.GetAccident(id)
	if err != nil {
		return err
	}
	if err := s.DB.Select("Photos").Delete(accident).Error; err != nil {
		return err
	}
	s.record(actor, "delete_accident", "accident", uintPtr(id), "")
	return nil
}

File: backend/app/services/accident_service.py
"""Accident service (Module 10)."""
from __future__ import annotations

from datetime import date
from typing import Any

from app.models import Accident, AccidentPhoto
from app.repositories import accident_repo, car_repo
from app.services import audit_service
from app.utils.errors import ValidationError


def _parse_date(value: Any) -> date:
    if isinstance(value, date):
        return value
    if not value:
        return date.today()
    return date.fromisoformat(str(value)[:10])


def create_accident(data: dict[str, Any], actor=None) -> Accident:
    car = car_repo.get_or_404(int(data["car_id"]))

    accident = Accident(
        car_id=car.id,
        client_id=int(data["client_id"]) if data.get("client_id") else None,
        rental_id=int(data["rental_id"]) if data.get("rental_id") else None,
        accident_date=_parse_date(data.get("accident_date")),
        description=data.get("description"),
        repair_cost=float(data.get("repair_cost") or 0),
    )
    for path in data.get("photos", []) or []:
        accident.photos.append(AccidentPhoto(file_path=path))

    accident_repo.add(accident)
    audit_service.record(
        "create_accident",
        user_id=getattr(actor, "id", None),
        username=getattr(actor, "username", None),
        entity="accident",
        entity_id=accident.id,
    )
    return accident


def update_accident(accident_id: int, data: dict[str, Any], actor=None) -> Accident:
    accident = accident_repo.get_or_404(accident_id)
    if "description" in data:
        accident.description = data["description"]
    if "repair_cost" in data and data["repair_cost"] is not None:
        accident.repair_cost = float(data["repair_cost"])
    if "accident_date" in data and data["accident_date"]:
        accident.accident_date = _parse_date(data["accident_date"])
    if "photos" in data and data["photos"] is not None:
        accident.photos.clear()
        for path in data["photos"]:
            accident.photos.append(AccidentPhoto(file_path=path))
    accident_repo.update(accident, {})
    return accident


def delete_accident(accident_id: int, actor=None) -> None:
    accident = accident_repo.get_or_404(accident_id)
    accident_repo.delete(accident)
    audit_service.record(
        "delete_accident",
        user_id=getattr(actor, "id", None),
        username=getattr(actor, "username", None),
        entity="accident",
        entity_id=accident_id,
    )
